test/replication-luatest/gh_6842_qsync_applier_order_test.lua

```diff
@@ -0,0 +1,370 @@
+        box.error.injection.set('ERRINJ_TX_DELAY_PRIO_ENDPOINT', 0.1)
+        -- Let them finish WAL writes.
+        box.error.injection.set('ERRINJ_RELAY_FASTER_THAN_TX', false)
+
+        local fiber = require('fiber')
+        for _, fid in pairs(fids) do
+            fiber.find(fid):join()
+        end
+        box.error.injection.set('ERRINJ_TX_DELAY_PRIO_ENDPOINT', 0)
+        return box.space.test:select()
+    end, {fids})
+    -- The txn was confirmed by another instance even before it was confirmed
+    -- locally.
+    luatest.assert_equals(rows, {{1}, {2}}, 'txn was confirmed')
+
+    local rows2 = g.server2:exec(function()
+        return box.space.test:select()
+    end)
+    luatest.assert_equals(rows2, rows, 'on instance 2 too')
+
+    -- The synchronous replication is functional - new owner can use the queue.
+    g.server2:exec(function()
+        box.space.test:replace{3}
+    end)
+    g.server1:wait_vclock_of(g.server2)
+    rows = g.server1:exec(function()
+        return box.space.test:select{}
+    end)
+    luatest.assert_equals(rows, {{1}, {2}, {3}}, 'synchro transactions work')
+end
+
+--
+-- Server 1 was a synchro queue owner. It starts a synchro txn. The txn is
+-- not written to WAL yet. Server 2 makes a PROMOTE which doesn't include that
+-- txn.
+--
+-- Then the PROMOTE is delivered to server 1 and goes to WAL too. Then the txn
+-- and PROMOTE WAL writes are processed by TX thread in a single batch without
+-- yields.
+--
+-- The bug was that the fiber authored the txn wasn't ready to a WAL write
+-- being success, txn being marked as non-synchro anymore (it happened on
+-- rollback in the synchro queue), but the txn signature being bad.
+--
+g.test_remote_promote_during_local_txn_not_including_it = function(g)
+    -- Start a synchro txn on server 1.
+    local fids = g.server1:exec(function()
+        box.ctl.promote()
+        local s = box.schema.create_space('test', {is_sync = true})
+        s:create_index('pk')
+        box.cfg{
+            replication_synchro_quorum = 3,
+        }
+        box.error.injection.set('ERRINJ_WAL_DELAY', true)
+        local fiber = require('fiber')
+        -- More than one transaction to ensure that it works not just for one.
+        local f1 = fiber.new(s.replace, s, {1})
+        f1:set_joinable(true)
+        local f2 = fiber.new(s.replace, s, {2})
+        f2:set_joinable(true)
+        return {f1:id(), f2:id()}
+    end)
+    -- Deliver server 1 promotion to 2. Otherwise server 2 might fail trying to
+    -- start its own promotion simultaneously.
+    g.server2:wait_vclock_of(g.server1)
+
+    -- Server 2 sends a PROMOTE not covering the txns to server 1.
+    g.server2:exec(function()
+        box.cfg{
+            replication_synchro_quorum = 1,
+        }
+        box.ctl.promote()
+    end)
+
+    -- Server 1 receives the PROMOTE.
+    wait_synchro_is_busy(g.server1)
+
+    local rows = g.server1:exec(function(fids)
```

Comment:
Same about this exec. It duplicates the one in the previous test.

Reply:
Not exactly. It operates on `ERRINJ_WAL_DELAY` while the other one operates on `ERRINJ_RELAY_FASTER_THAN_TX`. I tried now to extract at least the fiber spawn and join, but realized I would then need to make multiple execs, because we can't define functions on the storages properly. Didn't want to add more execs just for that.

Reply:
You could still leave one exec, see above.